File: visuals/src/Fable.Ripple.Dom/Hmr.js

import { substring } from "../../fable_modules/fable-library-js.5.13.0/String.js";
import { FSharpRef, Record } from "../../fable_modules/fable-library-js.5.13.0/Types.js";
import { record_type, class_type, int32_type, string_type } from "../../fable_modules/fable-library-js.5.13.0/Reflection.js";
import { addToSet, tryGetValue } from "../../fable_modules/fable-library-js.5.13.0/MapUtil.js";
import { disposeSafe, getEnumerator, equals, defaultOf, int32ToString } from "../../fable_modules/fable-library-js.5.13.0/Util.js";
import { filter, toArray } from "../../fable_modules/fable-library-js.5.13.0/Seq.js";
import { item as item_2 } from "../../fable_modules/fable-library-js.5.13.0/Array.js";
import { Var$1__set_Value_2B595, Var$1__get_Value, Var$1 } from "../Fable.Ripple/Types.js";
import { empty, cons, tail, head, isEmpty } from "../../fable_modules/fable-library-js.5.13.0/List.js";
import { Base_replaceTrackedNode, EmptyMarker } from "./Base.js";
import { Operators_IsNull } from "../../fable_modules/fable-library-js.5.13.0/FSharp.Core.js";
import { Var_create, Signal_onCleanup, Signal_effect, Signal_untracked, Signal_root } from "../Fable.Ripple/Api.js";

function key(url, name) {
    let matchValue;
    return (((matchValue = (url.indexOf("?") | 0), (matchValue === -1) ? url : substring(url, 0, matchValue))) + "#") + name;
}

const stash = new Map([]);

class Frame extends Record {
    constructor(Path, Counts, Seen) {
        super();
        this.Path = Path;
        this.Counts = Counts;
        this.Seen = Seen;
    }
}

function Frame_$reflection() {
    return record_type("Fable.Ripple.Dom.Hmr.Frame", [], Frame, () => [["Path", string_type], ["Counts", class_type("System.Collections.Generic.Dictionary`2", [string_type, int32_type])], ["Seen", class_type("System.Collections.Generic.HashSet`1", [string_type])]]);
}

function newFrame(path) {
    return new Frame(path, new Map([]), new Set([]));
}

let frame = newFrame("");

function place(k) {
    let n_1;
    let matchValue;
    let outArg = 0;
    matchValue = [tryGetValue(frame.Counts, k, new FSharpRef(() => (outArg | 0), (v) => {
        outArg = (v | 0);
    })), outArg];
    n_1 = (matchValue[0] ? matchValue[1] : 0);
    frame.Counts.set(k, n_1 + 1);
    return (((frame.Path + "/") + k) + "#") + int32ToString(n_1);
}

function enter(path) {
    const saved = frame;
    frame = newFrame(path);
    return saved;
}

function leave(saved) {
    let source, f1, objectArg;
    let matchValue;
    let outArg = defaultOf();
    matchValue = [tryGetValue(stash, frame.Path, new FSharpRef(() => outArg, (v) => {
        outArg = v;
    })), outArg];
    if (matchValue[0]) {
        const bag = matchValue[1];
        const stale = toArray((source = bag.keys(), filter((f1 = ((objectArg = frame.Seen, (item) => objectArg.has(item))), (arg) => !f1(arg)), source)));
        for (let idx = 0; idx <= (stale.length - 1); idx++) {
            const name = item_2(idx, stale);
            bag.delete(name);
        }
    }
    frame = saved;
}

/**
 * Keep a `Var` across a swap, keyed by the name it was bound to.
 * 
 * `create` runs only the first time. A type change between versions falls back to
 * a fresh Var rather than throwing.
 */
export function adopt(name, create) {
    addToSet(name, frame.Seen);
    let bag_2;
    let matchValue;
    let outArg = defaultOf();
    matchValue = [tryGetValue(stash, frame.Path, new FSharpRef(() => outArg, (v) => {
        outArg = v;
    })), outArg];
    if (matchValue[0]) {
        bag_2 = matchValue[1];
    }
    else {
        const bag_1 = new Map([]);
        stash.set(frame.Path, bag_1);
        bag_2 = bag_1;
    }
    let matchValue_1;
    let outArg_1 = defaultOf();
    matchValue_1 = [tryGetValue(bag_2, name, new FSharpRef(() => outArg_1, (v_1) => {
        outArg_1 = v_1;
    })), outArg_1];
    if (matchValue_1[0]) {
        const v_2 = matchValue_1[1];
        if (v_2 instanceof Var$1) {
            const typed = v_2;
            return typed;
        }
        else {
            const fresh = create();
            bag_2.set(name, fresh);
            return fresh;
        }
    }
    else {
        const fresh_1 = create();
        bag_2.set(name, fresh_1);
        return fresh_1;
    }
}

export function splice(run) {
    return (f => (f.__splice = true, f))((p, a) => run(p, a));
}

export function fragment(items) {
    return splice((parent, anchor) => {
        let rest = items;
        while (!isEmpty(rest)) {
            const matchValue = head(rest);
            if (matchValue instanceof Node) {
                parent.insertBefore(matchValue, anchor);
            }
            else {
                const element = parent;
                const item_1 = matchValue;
                if (item_1 instanceof Node) {
                    element.appendChild(item_1);
                }
                else if (item_1 instanceof EmptyMarker) {
                }
                else {
                    item_1(element);
                }
            }
            rest = tail(rest);
        }
    });
}

function capture(root) {
    const active = document.activeElement;
    if ((Operators_IsNull(active) ? true : Operators_IsNull(root)) ? true : !root.contains(active)) {
        return undefined;
    }
    else {
        const pathTo = (n_mut, acc_mut) => {
            pathTo:
            while (true) {
                const n = n_mut, acc = acc_mut;
                if (n === root) {
                    return acc;
                }
                else {
                    const matchValue = n.parentNode;
                    if (equals(matchValue, defaultOf())) {
                        return undefined;
                    }
                    else {
                        const p = matchValue;
                        let i = 0;
                        let c = p.firstChild;
                        while (!Operators_IsNull(c) && !(c === n)) {
                            i = ((i + 1) | 0);
                            c = c.nextSibling;
                        }
                        n_mut = p;
                        acc_mut = cons(i, acc);
                        continue pathTo;
                    }
                }
                break;
            }
        };
        const option_1 = pathTo(active, empty());
        if (option_1 != null) {
            return [option_1, active.selectionStart, active.selectionEnd];
        }
        else {
            return undefined;
        }
    }
}

function restore(root, captured) {
    const option_1 = captured;
    if (option_1 != null) {
        const tupledArg = option_1;
        const selStart = tupledArg[1];
        let node = root;
        let ok = !Operators_IsNull(root);
        const enumerator = getEnumerator(tupledArg[0]);
        try {
            while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
                const i = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]() | 0;
                if (ok) {
                    const children = node.childNodes;
                    if (i < children.length) {
                        node = children.item(i);
                    }
                    else {
                        ok = false;
                    }
                }
            }
        }
        finally {
            disposeSafe(enumerator);
        }
        if (ok && !Operators_IsNull(node)) {
            const el = node;
            if (!Operators_IsNull(el.focus)) {
                el.focus();
                if (!Operators_IsNull(selStart)) {
                    try {
                        el.setSelectionRange(selStart, tupledArg[2]);
                    }
                    catch (matchValue) {
                    }
                }
            }
        }
    }
}

function boundaryOf(k, impl, args) {
    const path = place(k);
    const build = (f) => Signal_root(() => {
        const saved = enter(path);
        const item = Signal_untracked(() => (f.apply(null, args)));
        leave(saved);
        return item;
    });
    const patternInput = build(Var$1__get_Value(impl));
    const firstItem = patternInput[0];
    const firstDispose = patternInput[1];
    if (firstItem instanceof Node) {
        let current = firstItem;
        let dispose = firstDispose;
        let built = true;
        Signal_effect(() => {
            const f_1 = Var$1__get_Value(impl);
            if (built) {
                built = false;
            }
            else {
                globalThis.__rebuilds = (globalThis.__rebuilds || 0) + 1;
                const t0 = performance.now();
                const parent = current.parentNode;
                const captured = capture(current);
                disposeSafe(dispose);
                const patternInput_1 = build(f_1);
                const next = patternInput_1[0];
                dispose = patternInput_1[1];
                if (next instanceof Node) {
                    const fresh = next;
                    if (!Operators_IsNull(parent)) {
                        parent.replaceChild(fresh, current);
                        Base_replaceTrackedNode(current, fresh);
                    }
                    current = fresh;
                    restore(current, captured);
                }
                (globalThis.__rebuildMs = globalThis.__rebuildMs || []).push((performance.now()) - t0);
            }
        });
        Signal_onCleanup(() => {
            disposeSafe(dispose);
        });
        return current;
    }
    else {
        disposeSafe(firstDispose);
        return (parent_1) => {
            const startA = document.createComment("b[");
            const endA = document.createComment("]b");
            parent_1.appendChild(startA);
            parent_1.appendChild(endA);
            let current_1 = undefined;
            const clear = () => {
                const option_1 = current_1;
                if (option_1 != null) {
                    disposeSafe(option_1);
                }
                current_1 = undefined;
                let n = startA.nextSibling;
                while (!Operators_IsNull(n) && !(n === endA)) {
                    const next_1 = n.nextSibling;
                    parent_1.removeChild(n);
                    n = next_1;
                }
            };
            Signal_effect(() => {
                const f_2 = Var$1__get_Value(impl);
                globalThis.__rebuilds = (globalThis.__rebuilds || 0) + 1;
                const captured_1 = capture(parent_1);
                const t0_1 = performance.now();
                clear();
                const patternInput_2 = build(f_2);
                const item_1 = patternInput_2[0];
                if (item_1 instanceof EmptyMarker) {
                }
                else if (typeof item_1 === "function") {
                    if (item_1 != null && item_1.__splice === true) {
                        item_1(parent_1, endA);
                    }
                    else {
                        item_1(parent_1);
                    }
                }
                else {
                    parent_1.insertBefore(item_1, endA);
                }
                current_1 = patternInput_2[1];
                restore(parent_1, captured_1);
                (globalThis.__rebuildMs = globalThis.__rebuildMs || []).push((performance.now()) - t0_1);
            });
            Signal_onCleanup(clear);
        };
    }
}

export function define(url, name, fn) {
    const k = key(url, name);
    let impl;
    let matchValue;
    let outArg = defaultOf();
    matchValue = [tryGetValue((globalThis.__RIPPLE_HMR__ || (globalThis.__RIPPLE_HMR__ = new Map())), k, new FSharpRef(() => outArg, (v) => {
        outArg = v;
    })), outArg];
    if (matchValue[0]) {
        const v_1 = matchValue[1];
        Var$1__set_Value_2B595(v_1, fn);
        impl = v_1;
    }
    else {
        const v_2 = Var_create(fn);
        ((globalThis.__RIPPLE_HMR__ || (globalThis.__RIPPLE_HMR__ = new Map()))).set(k, v_2);
        impl = v_2;
    }
    return (...a) => (boundaryOf)(k, impl, a);
}

File: src/Fable.Ripple.Dom/Hmr.fs
namespace Fable.Ripple.Dom

open System
open System.Collections.Generic
open Browser
open Browser.Types
open Fable.Core
open Fable.Core.JsInterop
open Fable.Ripple
open Fable.Ripple.Dom.Base

/// The runtime behind `[<Component>]`. Called only from code the plugin emits, and
/// the plugin emits none of it outside a Debug build.
module Hmr =

    let private key (url: string) (name: string) =
        let url =
            match url.IndexOf "?" with
            | -1 -> url
            | i -> url.Substring(0, i)

        url + "#" + name

    /// On globalThis, not in this module: this module can itself be hot-replaced.
    [<Emit("(globalThis.__RIPPLE_HMR__ || (globalThis.__RIPPLE_HMR__ = new Map()))")>]
    let private store: Dictionary<string, Var<obj>> = jsNative

    // --- state carried across a swap --------------------------------------------

    let private stash = Dictionary<string, Dictionary<string, obj>>()

    /// Where the boundary being built sits, and the tally for the ones it builds.
    type private Frame =
        {
            Path: string
            Counts: Dictionary<string, int>
            Seen: HashSet<string>
        }

    let private newFrame (path: string) =
        {
            Path = path
            Counts = Dictionary()
            Seen = HashSet()
        }

    let mutable private frame = newFrame ""

    /// The place a boundary takes among its siblings, fixed when it is built.
    ///
    /// Counted per component, so instances of one do not renumber instances of another.
    let private place (k: string) =
        let n =
            match frame.Counts.TryGetValue k with
            | true, n -> n
            | _ -> 0

        frame.Counts.[k] <- n + 1
        frame.Path + "/" + k + "#" + string n

    let private enter (path: string) =
        let saved = frame
        frame <- newFrame path
        saved

    let private leave (saved: Frame) =
        match stash.TryGetValue frame.Path with
        | true, bag ->
            let stale = bag.Keys |> Seq.filter (frame.Seen.Contains >> not) |> Seq.toArray

            for name in stale do
                bag.Remove name |> ignore
        | _ -> ()

        frame <- saved

    /// Keep a `Var` across a swap, keyed by the name it was bound to.
    ///
    /// `create` runs only the first time. A type change between versions falls back to
    /// a fresh Var rather than throwing.
    let adopt (name: string) (create: unit -> Var<'T>) : Var<'T> =
        frame.Seen.Add name |> ignore

        let bag =
            match stash.TryGetValue frame.Path with
            | true, bag -> bag
            | _ ->
                let bag = Dictionary<string, obj>()
                stash.[frame.Path] <- bag
                bag

        match bag.TryGetValue name with
        | true, v ->
            match v with
            | :? Var<'T> as typed -> typed
            | _ ->
                let fresh = create ()
                bag.[name] <- box fresh
                fresh
        | _ ->
            let fresh = create ()
            bag.[name] <- box fresh
            fresh

    // --- the boundary ------------------------------------------------------------

    [<Emit("(f => (f.__splice = true, f))((p, a) => $0(p, a))")>]
    let private spliceFn (run: Node -> Node -> unit) : Element -> unit = jsNative

    let splice (run: Node -> Node -> unit) : DomItem = Apply(spliceFn run)

    [<Emit("$0 != null && $0.__splice === true")>]
    let private isSplice (item: DomItem) : bool = jsNative

    [<Emit("$0($1, $2)")>]
    let private runSplice (item: DomItem) (parent: Node) (anchor: Node) : unit = jsNative

    [<Emit("globalThis.__rebuilds = (globalThis.__rebuilds || 0) + 1")>]
    let private countRebuild () : unit = jsNative

    [<Emit("performance.now()")>]
    let private now () : float = jsNative

    [<Emit("(globalThis.__rebuildMs = globalThis.__rebuildMs || []).push($0)")>]
    let private recordRebuild (ms: float) : unit = jsNative

    let fragment (items: DomItem list) : DomItem =
        splice (fun parent anchor ->
            let mutable rest = items

            while not (List.isEmpty rest) do
                match List.head rest with
                | Child node -> parent.insertBefore (node, anchor) |> ignore
                | item -> applyItem (parent :?> Element) item

                rest <- List.tail rest
        )

    [<Emit("$0.apply(null, $1)")>]
    let private applyArgs (fn: obj) (args: obj[]) : DomItem = jsNative

    // The focused element is located by its position under `root`, so an edit that
    // changes the shape around it loses focus.

    let private capture (root: Node) =
        let active = document.activeElement

        if isNull (box active) || isNull (box root) || not (root.contains active) then
            None
        else
            let rec pathTo (n: Node) acc =
                if obj.ReferenceEquals(n, root) then
                    Some acc
                else
                    match n.parentNode with
                    | null -> None
                    | p ->
                        let mutable i = 0
                        let mutable c = p.firstChild

                        while not (isNull c) && not (obj.ReferenceEquals(c, n)) do
                            i <- i + 1
                            c <- c.nextSibling

                        pathTo p (i :: acc)

            pathTo (active :> Node) []
            |> Option.map (fun p -> p, active?selectionStart, active?selectionEnd)

    let private restore (root: Node) (captured: (int list * obj * obj) option) =
        captured
        |> Option.iter (fun (path, selStart, selEnd) ->
            let mutable node = root
            let mutable ok = not (isNull (box root))

            for i in path do
                if ok then
                    let children = node.childNodes

                    if i < children.length then
                        node <- children.item i
                    else
                        ok <- false

            if ok && not (isNull (box node)) then
                let el: HTMLElement = unbox node

                if not (isNull (box el?focus)) then
                    el.focus ()

                    if not (isNull selStart) then
                        try
                            el?setSelectionRange (selStart, selEnd)
                        with _ ->
                            ()
        )

    let private boundaryOf (k: string) (impl: Var<obj>) (args: obj[]) : DomItem =
        let path = place k

        let build (f: obj) =
            Signal.root (fun () ->
                let saved = enter path
                let item = Signal.untracked (fun () -> applyArgs f args)
                leave saved
                item
            )

        let firstItem, firstDispose = build impl.Value

        match firstItem with
        // A component that renders one element is that element, so it can sit anywhere
        // a `DomItem` can - an `Html.each` row included, which needs one node per key.
        | Child node ->
            let mutable current = node
            let mutable dispose = firstDispose
            let mutable built = true

            Signal.effect (fun () ->
                let f = impl.Value

                if built then
                    built <- false
                else
                    countRebuild ()
                    let t0 = now ()
                    let parent = current.parentNode
                    let captured = capture current
                    dispose.Dispose()

                    let next, d = build f
                    dispose <- d

                    match next with
                    | Child fresh ->
                        if not (isNull parent) then
                            parent.replaceChild (fresh, current) |> ignore
                            replaceTrackedNode current fresh

                        current <- fresh
                        restore current captured
                    | _ -> ()

                    recordRebuild (now () - t0)
            )
            |> ignore

            Signal.onCleanup (fun () -> dispose.Dispose())
            Child current

        // A fragment, bare attributes or nothing at all has no single node to stand
        // for it, so it is rebuilt between anchors in its parent.
        | _ ->
            firstDispose.Dispose()

            Apply(fun parent ->
                let startA = document.createComment "b[" :> Node
                let endA = document.createComment "]b" :> Node
                parent.appendChild startA |> ignore
                parent.appendChild endA |> ignore

                let mutable current: IDisposable option = None

                let clear () =
                    current |> Option.iter (fun d -> d.Dispose())
                    current <- None
                    let mutable n = startA.nextSibling

                    while not (isNull n) && not (obj.ReferenceEquals(n, endA)) do
                        let next = n.nextSibling
                        parent.removeChild n |> ignore
                        n <- next

                Signal.effect (fun () ->
                    let f = impl.Value
                    countRebuild ()
                    let captured = capture parent
                    let t0 = now ()
                    clear ()

                    let item, dispose = build f

                    match item with
                    | Child node -> parent.insertBefore (node, endA) |> ignore
                    | Empty _ -> ()
                    | Apply run ->
                        if isSplice item then
                            runSplice item parent endA
                        else
                            run parent

                    current <- Some dispose

                    restore parent captured
                    recordRebuild (now () - t0)
                )
                |> ignore

                Signal.onCleanup clear
            )

    /// Emitted by the plugin in place of the component's own body.
    ///
    /// Always swaps: a re-executed module closes over the new imports, so keeping the
    /// old implementation serves stale values. The wrapper is variadic, so the boundary
    /// can re-invoke with the arguments the call site passed.
    [<Emit("(...a) => $0($1, $2, a)")>]
    let private variadic
        (build: string -> Var<obj> -> obj[] -> DomItem)
        (k: string)
        (impl: Var<obj>)
        : obj
        =
        jsNative

    let define (url: string) (name: string) (fn: obj) : obj =
        let k = key url name

        let impl =
            match store.TryGetValue k with
            | true, v ->
                v.Value <- fn
                v
            | _ ->
                let v = Var.create fn
                store.[k] <- v
                v

        variadic boundaryOf k impl
